### ISSUE TITLE
Types: `plugin` and `parser` types incompatible with ESLint types

### ISSUE THREAD

**Issue by JstnMcBrd** [typescript-eslint/typescript-eslint#9110]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

I am trying to create a type-safe ESLint config using `@typescript-eslint`. But it appears the types are not formatted correctly.

Here is a simplified version of what I have:
```ts
import tsESLint from 'typescript-eslint';
import type { Linter } from 'eslint';

export function config(): Linter.FlatConfig[] {
    return [
        {
            plugins: {
                // Throws type error
                // The type of 'TSESLint.FlatConfig.Plugin' is not compatible with 'ESLint.Plugin'
                '@typescript-eslint': tsESLint.plugin,
            },
            languageOptions: {
                // Throws type error
                // The type of 'TSESLint.Parser.LooseParserModule' is not compatible with 'Linter.FlatConfigParserModule'
                parser: tsESLint.parser,
            },
            // ...
        },
    ];
}
```

Here's the error message for the plugin:
```
Type 'import(".../@typescript-eslint/utils/dist/ts-eslint/Config").FlatConfig.Plugin' is not assignable to type 'import("c:/Users/jstnm/OneDrive/Programming/eslint-config/node_modules/@types/eslint/index").ESLint.Plugin'.
  Types of property 'configs' are incompatible.
    Type 'SharedConfigs | undefined' is not assignable to type 'Record<string, FlatConfig<RulesRecord> | FlatConfig<RulesRecord>[] | ConfigData<RulesRecord>> | undefined'.
      Type 'SharedConfigs' is not assignable to type 'Record<string, FlatConfig<RulesRecord> | FlatConfig<RulesRecord>[] | ConfigData<RulesRecord>>'.
        'string' index signatures are incompatible.
          Type 'Config' is not assignable to type 'FlatConfig<RulesRecord> | FlatConfig<RulesRecord>[] | ConfigData<RulesRecord>'.
            Type 'Config' is not assignable to type 'FlatConfig<RulesRecord>'.
              Types of property 'files' are incompatible.
                Type '(FileSpec | FileSpec[])[] | undefined' is not assignable to type '(string | string[])[] | undefined'.
                  Type '(FileSpec | FileSpec[])[]' is not assignable to type '(string | string[])[]'.
                    Type 'FileSpec | FileSpec[]' is not assignable to type 'string | string[]'.
                      Type '(filePath: string) => boolean' is not assignable to type 'string | string[]'.
```

Here's the error message for the parser (slightly different between v7 and v8):
```
// typescript-eslint@7.9.0
Type 'LooseParserModule' is not assignable to type 'FlatConfigParserModule | undefined'.
  Type '{ meta?: { name?: string | undefined; version?: string | undefined; } | undefined; parseForESLint(text: string, options?: unknown): { ast: unknown; services?: unknown; scopeManager?: unknown; visitorKeys?: unknown; }; }' is not assignable to type 'FlatConfigParserModule | undefined'.
    Type '{ meta?: { name?: string | undefined; version?: string | undefined; } | undefined; parseForESLint(text: string, options?: unknown): { ast: unknown; services?: unknown; scopeManager?: unknown; visitorKeys?: unknown; }; }' is not assignable to type 'Omit<ParserModule, "parseForESLint"> & { parseForESLint(text: string, options?: any): Omit<ESLintParseResult, "ast" | "scopeManager"> & { ...; }; }'.
      Type '{ meta?: { name?: string | undefined; version?: string | undefined; } | undefined; parseForESLint(text: string, options?: unknown): { ast: unknown; services?: unknown; scopeManager?: unknown; visitorKeys?: unknown; }; }' is not assignable to type '{ parseForESLint(text: string, options?: any): Omit<ESLintParseResult, "ast" | "scopeManager"> & { ast: unknown; scopeManager?: unknown; }; }'.
        The types returned by 'parseForESLint(...)' are incompatible between these types.
          Type '{ ast: unknown; services?: unknown; scopeManager?: unknown; visitorKeys?: unknown; }' is not assignable to type 'Omit<ESLintParseResult, "ast" | "scopeManager"> & { ast: unknown; scopeManager?: unknown; }'.
            Type '{ ast: unknown; services?: unknown; scopeManager?: unknown; visitorKeys?: unknown; }' is not assignable to type 'Omit<ESLintParseResult, "ast" | "scopeManager">'.
              Types of property 'visitorKeys' are incompatible.
                Type 'unknown' is not assignable to type 'VisitorKeys | undefined'.

// typescript-eslint@8.0.0-alpha.13
Type '{ meta?: { name?: string; version?: string; }; parseForESLint(text: string, options?: unknown): { ast: unknown; services?: unknown; scopeManager?: unknown; visitorKeys?: unknown; }; }' is not assignable to type 'FlatConfigParserModule'.
  Type '{ meta?: { name?: string; version?: string; }; parseForESLint(text: string, options?: unknown): { ast: unknown; services?: unknown; scopeManager?: unknown; visitorKeys?: unknown; }; }' is not assignable to type 'Omit<ParserModule, "parseForESLint"> & { parseForESLint(text: string, options?: any): Omit<ESLintParseResult, "ast" | "scopeManager"> & { ...; }; }'.
    Type '{ meta?: { name?: string; version?: string; }; parseForESLint(text: string, options?: unknown): { ast: unknown; services?: unknown; scopeManager?: unknown; visitorKeys?: unknown; }; }' is not assignable to type '{ parseForESLint(text: string, options?: any): Omit<ESLintParseResult, "ast" | "scopeManager"> & { ast: unknown; scopeManager?: unknown; }; }'.
      The types returned by 'parseForESLint(...)' are incompatible between these types.
        Type '{ ast: unknown; services?: unknown; scopeManager?: unknown; visitorKeys?: unknown; }' is not assignable to type 'Omit<ESLintParseResult, "ast" | "scopeManager"> & { ast: unknown; scopeManager?: unknown; }'.
          Type '{ ast: unknown; services?: unknown; scopeManager?: unknown; visitorKeys?: unknown; }' is not assignable to type 'Omit<ESLintParseResult, "ast" | "scopeManager">'.
            Types of property 'visitorKeys' are incompatible.
              Type 'unknown' is not assignable to type 'VisitorKeys'.
                Index signature for type 'string' is missing in type '{}'.
```

### Cause

`typescript-eslint` is probably not used this way very often. The types are being held to a much higher standard of accuracy by requiring that they match ESLint's types. It appears there some are mistakes.

### Workarounds

The `plugin` and `parser` can be force-typecasted to the correct type, but it is unsafe and not ideal.

### Reproduction Repository Link

https://github.com/JstnMcBrd/typescript-eslint-repro

### Repro Steps

1) clone the repo
2) `npm install`
     optionally, change versions with `npm install typescript-eslint@8.0.0-alpha.13`
3) `npm run build`

### Versions

| package                                | versions |
| -------------------------------------- | ------- |
| `typescript-eslint`                    | `7.9.0` (latest), `8.0.0-alpha.13` (next) |
| `TypeScript`                           | `5.4.5` (latest) |
| `node`                                 | `20.13.1` (LTS) |

**Comment by auvred** [comment]
> ```ts
>             plugins: {
>                 // Throws type error
>                 // The type of 'TSESLint.FlatConfig.Plugin' is not compatible with 'ESLint.Plugin'
>                 '@typescript-eslint': tsESLint.plugin,
>             },
> ```

Yup, thanks for reporting! This issue has interesting story:

1. typescript-eslint didn't allow to pass `(filePath: string) => boolean` to `files` property - https://github.com/typescript-eslint/typescript-eslint/issues/8467
1. temporary stubbed on typescript-eslint side - https://github.com/typescript-eslint/typescript-eslint/pull/8460
  https://github.com/typescript-eslint/typescript-eslint/blob/165d8f7a0de7fefdc16658c1161773b1bd5f3a2a/packages/utils/src/ts-eslint/Config.ts#L235-L243
1. reported on ESLint side, agreed to remove `(filePath: string) => boolean` from `@types/eslint` - https://github.com/eslint/eslint/issues/18118
1. removed `(filePath: string) => boolean` from `@types/eslint`- https://github.com/DefinitelyTyped/DefinitelyTyped/pull/69210

So it's time to finish this journey and remove `(filePath: string) => boolean` added in step (2) from our types!

Although I'm not sure whether to consider this as a bug fix or a breaking change?

cc @bradzacher 

--- 

> ```ts
>             languageOptions: {
>                 // Throws type error
>                 // The type of 'TSESLint.Parser.LooseParserModule' is not compatible with 'Linter.FlatConfigParserModule'
>                 parser: tsESLint.parser,
>             },
> ```

Looks like a duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/8613

**Comment by bradzacher** [comment]
No it's a fix - that signature was only added to keep the ESLint types happy and it's not used by anyone.

It was only ever added due to the original flat config author for `@types/eslint` reading the implementation of the flat config package. It's never been a documented thing and only existed so flat configs could handle some obscure eslintrc feature.

But it was never part of the public API - hence it was removed. So yeah we delete it too and close the loop on this silly issue!

**Comment by JstnMcBrd** [comment]
Thanks for the quick fix!

-----

About the second type error (@bradzacher):

I see on #8613 that it was closed as "not planned" because "we purposely do not seek compatibility with `@types/eslint`". I'm not sure I understand why.

Sure, your parser is going to be more complicated that the `Parser` from `@types/eslint` - "their types are not designed to take into account our AST and such" - but at the API level, shouldn't it still be compatible with the parser type that ESLint is expecting? Shouldn't it at least extend the `Parser`, rather than being incompatible with it?

It feels unsafe to pass a parser into my config that is explicitly incompatible with the parser API that ESLint is expecting. And if it is compatible, then the types should reflect that, no?

**Comment by bradzacher** [comment]
Type compatibility != runtime compatibility.
We are the parser that allows you to use typescript with eslint and we are 110% compatible at runtime.

As I explained https://github.com/typescript-eslint/typescript-eslint/issues/8613#issuecomment-1984076953 - our types are designed to be permissive so you have a good DevX, but `@types/eslint`'s types are designed to exactly describe what is allowed.

Here is a simplified example of what I mean:
```ts
// typescript-eslint
type LooseParser = string;
type Parser = 'value';
export const parser: Parser;
export type Config = { parser?: LooseParser };

// @types/eslint
type Parser = 'value';
export const parser: Parser;
export type Config = { parser?: Parser };
```

Our parser types are strictly defined to match the runtime types - same as `@types/eslint`.
However unlike `@types/eslint` our config type is intentionally defined loosely. This widens the things you're allowed to pass to your config.
But why? Well to account for differences in types that packages may define.
In reality ESLint is much more permissive than what `@types/eslint` defines - for example eslint allows any AST as long as it follows a very, very basic structure - but a lot of `@types/eslint` only accepts the JS/estree AST. Until recently that was the only ast you could use with the config (I had to push hard to get them to loosen that type to allow our AST to be assigned).

To clearly state our intent: our types are designed to be as permissive as possible so that you can assign anything valid from the ecosystem. They are not designed to be compatible with the overly strict `@types/eslint`.
Our recommendation for typing your flat config is the [`config` function defined by our flat config package](https://typescript-eslint.io/packages/typescript-eslint#config)

**Comment by JstnMcBrd** [comment]
> In reality ESLint is much more permissive than what @types/eslint defines

Hmm, then it sounds like the problem is with `@types/eslint`. I still firmly believe that if your parser is compatible with ESLint, then the types should be compatible too. Types should correctly define what is allowed and what is not. If the ESLint types are incompatible with things that ESLint permits, then those types are inaccurate. But I can take that up with them instead. Thank you for the explanation.

**Comment by bradzacher** [comment]
Look at it like this.

Our types enforce best practices - for example we enforce certain rule properties are non-optional so that people that use our tooling are forced to follow those practices (like properly documenting their rule).

But `@types/eslint` is built with a different intention - to add types that exhaustively show what may or may not be possible.

So in that example a rule defined with `@types/eslint` is not assignable to our rule type (because optional properties are not assignable to required ones). From that it follows that a plugin defined with `@types/eslint` is not assignable to our plugin type.

We can't have that though - because some parts of the ecosystem have chosen to use `@types/eslint` to declare their rules/plugin!

Which is why we define these "loose" types - to widen the set of what can be provided to the config!

Our config types focus on the developer experience, not on strict and granular correctness. We aim to provide enough of a shape to the types to allow us to validate that your passing the right thing without deeply validating the types. Eg the types we declare will ensure you don't pass a parser to the place expecting a plugin - but it won't validate that the parser function's return type's `visitorKeys` properties is exactly a `Record<string, string[]>`.
